File: answer/4_strike/3_refactor/bowling_game.py
class Game:
    """볼링 게임 클래스"""

    def __init__(self):
        """
        새로운 게임을 초기화합니다.
        """
        self.rolls = []

    def roll(self, pins: int) -> None:
        """
        플레이어가 공을 굴려 쓰러뜨린 핀의 개수를 기록합니다.

        Args:
            pins: 쓰러뜨린 핀의 수 (0-10)
        """
        self.rolls.append(pins)

    def score(self) -> int:
        total = 0
        roll_index = 0
        
        for frame in range(10):
            if self._is_strike(roll_index):
                total += 10 + self._strike_bonus(roll_index)
                roll_index += 1
            elif self._is_spare(roll_index):
                total += 10 + self._spare_bonus(roll_index)
                roll_index += 2
            else:
                total += self._frame_score(roll_index)
                roll_index += 2
        
        return total

    def _is_strike(self, roll_index: int) -> bool:
        """스트라이크 판정: 첫 투구가 10핀"""
        return self.rolls[roll_index] == 10

    def _strike_bonus(self, roll_index: int) -> int:
        """스트라이크 보너스: 다음 2투구"""
        return self.rolls[roll_index + 1] + self.rolls[roll_index + 2]

    def _is_spare(self, roll_index: int) -> bool:
        """스페어 판정을 별도 메서드로 추출"""
        return self.rolls[roll_index] + self.rolls[roll_index + 1] == 10

    def _spare_bonus(self, roll_index: int) -> int:
        """스페어 보너스 계산을 별도 메서드로 추출"""
        return self.rolls[roll_index + 2]

    def _frame_score(self, roll_index: int) -> int:
        """일반 프레임 점수 계산을 별도 메서드로 추출"""
        return self.rolls[roll_index] + self.rolls[roll_index + 1]

File: answer/4_strike/3_refactor/bowlingGame.js
/**
 * 볼링 게임 클래스
 */
export class Game {
    /**
     * 새로운 게임을 초기화합니다.
     */
    constructor() {
        this.rolls = [];
    }

    /**
     * 플레이어가 공을 굴려 쓰러뜨린 핀의 개수를 기록합니다.
     *
     * @param {number} pins - 쓰러뜨린 핀의 수 (0-10)
     */
    roll(pins) {
        this.rolls.push(pins);
    }

    score() {
        let total = 0;
        let rollIndex = 0;

        for (let frame = 0; frame < 10; frame++) {
            if (this._isStrike(rollIndex)) {
                total += 10 + this._strikeBonus(rollIndex);
                rollIndex += 1;
            } else if (this._isSpare(rollIndex)) {
                total += 10 + this._spareBonus(rollIndex);
                rollIndex += 2;
            } else {
                total += this._frameScore(rollIndex);
                rollIndex += 2;
            }
        }

        return total;
    }

    /**
     * 스트라이크 판정: 첫 투구가 10핀
     */
    _isStrike(rollIndex) {
        return this.rolls[rollIndex] === 10;
    }

    /**
     * 스트라이크 보너스: 다음 2투구
     */
    _strikeBonus(rollIndex) {
        return this.rolls[rollIndex + 1] + this.rolls[rollIndex + 2];
    }

    /**
     * 스페어 판정을 별도 메서드로 추출
     */
    _isSpare(rollIndex) {
        return this.rolls[rollIndex] + this.rolls[rollIndex + 1] === 10;
    }

    /**
     * 스페어 보너스 계산을 별도 메서드로 추출
     */
    _spareBonus(rollIndex) {
        return this.rolls[rollIndex + 2];
    }

    /**
     * 일반 프레임 점수 계산을 별도 메서드로 추출
     */
    _frameScore(rollIndex) {
        return this.rolls[rollIndex] + this.rolls[rollIndex + 1];
    }
}
